pkg/build/builder/common_test.go

```diff
@@ -113,3 +121,381 @@ func TestContainerName(t *testing.T) {
+			},
+			want: want{
+				Out: heredoc.Doc(`
+				FROM scratch as test
+				COPY --from=test:latest /a /b
+				`),
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM scratch as test
+				COPY --from=test:latest /a /b
+				`),
+			build: []buildapiv1.ImageSource{
+				{From: corev1.ObjectReference{Kind: "DockerImage", Name: "nginx:latest"}, As: []string{"test"}},
+			},
+			want: want{
+				Out: heredoc.Doc(`
+				FROM scratch as test
+				COPY --from=test:latest /a /b
+				`),
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM scratch as test
+				FROM other
+				COPY --from=test /a /b
+				`),
+			build: []buildapiv1.ImageSource{
+				{From: corev1.ObjectReference{Kind: "DockerImage", Name: "nginx:latest"}, As: []string{"test"}},
+			},
+			want: want{
+				Out: heredoc.Doc(`
+				FROM scratch as test
+				FROM other
+				COPY --from=test /a /b
+				`),
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM scratch as test
+				COPY --from=test:latest /a /b
+				`),
+			build: []buildapiv1.ImageSource{
+				{From: corev1.ObjectReference{Kind: "DockerImage", Name: "nginx:latest"}, As: []string{"test:latest"}},
+			},
+			want: want{
+				Out: heredoc.Doc(`
+				FROM scratch as test
+				COPY --from=nginx:latest /a /b
+				`),
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM scratch as test
+				COPY --from=test:latest /a /b
+				`),
+			from: &corev1.ObjectReference{
+				Kind: "DockerImage",
+				Name: "from-image:v1",
+			},
+			build: []buildapiv1.ImageSource{
+				{From: corev1.ObjectReference{Kind: "DockerImage", Name: "nginx:latest"}, As: []string{"scratch", "test:latest"}},
+			},
+			want: want{
+				Out: heredoc.Doc(`
+				FROM from-image:v1 as test
+				COPY --from=nginx:latest /a /b
+				`),
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM scratch as test
+				COPY --from=test:latest /a /b
+				`),
+			from: &corev1.ObjectReference{
+				Kind: "DockerImage",
+				Name: "from-image:v1",
+			},
+			build: []buildapiv1.ImageSource{
+				{From: corev1.ObjectReference{Kind: "DockerImage", Name: "nginx:latest"}, As: []string{"scratch", "test:latest", "from-image:v1"}},
+			},
+			want: want{
+				Out: heredoc.Doc(`
+				FROM nginx:latest as test
+				COPY --from=nginx:latest /a /b
+				`),
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM scratch as test
+				COPY --from=test:latest /a /b
+				`),
+			from: &corev1.ObjectReference{
+				Kind: "ImageStreamTag",
+				Name: "from-image:v1",
+			},
+			want: want{
+				Out: heredoc.Doc(`
+				FROM scratch as test
+				COPY --from=test:latest /a /b
+				`),
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM test
+				FROM scratch as test
+				COPY --from=test /a /b
+				`),
+			build: []buildapiv1.ImageSource{
+				{From: corev1.ObjectReference{Kind: "DockerImage", Name: "nginx:latest"}, As: []string{"test", "scratch"}},
+			},
+			want: want{
+				Out: heredoc.Doc(`
+				FROM nginx:latest
+				FROM nginx:latest as test
+				COPY --from=test /a /b
```

Comment:
what's the behavior if this COPY comes *before* the second FROM that defines "as test"? (needs a test regardless)

Reply:
Tries to pull test

Reply:
not seeing a test added to cover this?   It looks like you added one for the findReferencedImages, which is half the battle but i'd still want to see one for the substitution also.  Especially since a case like:

```
FROM other
COPY --from=test /a /b
FROM scratch as test
COPY --from=test /a /b
```


Needs to come out as:
```
FROM <othersubstitute>
COPY --from=test<unchanged> /a /b
FROM <scratchsubstitute> as test
COPY --from=<testsubstitute> /a /b
```

(or so i'd expect?)

Reply:
The last test covers it.

Reply:
sorry, missed that for some reason.  maybe looking at older commits.